### PR TITLE
Add check to ensure error code explanations are not removed anymore even if not emitted

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
The error explanations are useful in case you use older version of the compiler. Even more if they had an explanation. If they are not emitted, their explanations should be updated but not removed (as we did for a few of them, like E0001).

r? @Mark-Simulacrum

### CODE REVIEW

File: src/ci/scripts/should-skip-this.sh [hunk 1]
```diff
@@ -18,7 +18,11 @@ BASE_COMMIT="$(git merge-base FETCH_HEAD HEAD)"
 
 echo "Searching for toolstate changes between $BASE_COMMIT and $(git rev-parse HEAD)"
 
-if git diff "$BASE_COMMIT" | grep --quiet "^index .* 160000"; then
+if git log HEAD..."$BASE_COMMIT" | grep --quiet "compiler/rustc_error_codes/src/error_codes/"; then
```
Comment (nagisa) [inline src/ci/scripts/should-skip-this.sh:21]: Hm, how does this work?

This command fails over here even if no error codes have been touched. The `git log HEAD..."$BASE_COMMIT"` does not output a list of changed files or anything of the sort here. Furthermore, even if it did, I don't see how this checks for deletions specifically, as opposed to any changes to that directory.

Comment (GuillaumeGomez) [reply src/ci/scripts/should-skip-this.sh:21]: You're absolutely right, I badly pasted the command and didn't even pay attention... The real command is:

```
git log HEAD..."$BASE_COMMIT" | --diff-filter=D --summary | grep delete | grep --quiet "compiler/rustc_error_codes/src/error_codes/"
```

File: src/ci/scripts/should-skip-this.sh [hunk 2]
```diff
@@ -18,7 +18,12 @@ BASE_COMMIT="$(git merge-base FETCH_HEAD HEAD)"
 
 echo "Searching for toolstate changes between $BASE_COMMIT and $(git rev-parse HEAD)"
 
-if git diff "$BASE_COMMIT" | grep --quiet "^index .* 160000"; then
+if (git log HEAD..."$BASE_COMMIT" --diff-filter=D --summary | grep delete \
+        | grep --quiet "compiler/rustc_error_codes/src/error_codes/"); then
+    echo "Error code explanations should never be removed!"
+    echo "Take a look at E0001 to see how to handle it."
+    exit 1
```
Comment (Mark-Simulacrum) [inline src/ci/scripts/should-skip-this.sh:25]: Why is this check added to a skip script? This doesn't seem like the right place to add this.

I would probably try moving this to a script like validate-toolstate.sh in mingw-check, and hooking it up to run there.

Comment (GuillaumeGomez) [reply src/ci/scripts/should-skip-this.sh:25]: I simply didn't know where to put it. Now I know, thanks for the indication! :)

**Comment by GuillaumeGomez** [comment]
Moved the check into `validate-toolstate.sh` in mingw-check.

**Comment by Mark-Simulacrum** [comment]
This is unrelated to validating toolstate, so it should go in a separate file.

**Comment by GuillaumeGomez** [comment]
Created a new bash script I'm now calling in the `Dockerfile`.

**Comment by Mark-Simulacrum** [comment]
Please push a temp commit to make sure this works (mingw-check should fail on the PR) and also we may need to somehow skip this script in local docker invocations (where FETCH_HEAD is meaningless). I'm not sure what the best way of doing that would be, but it may also be OK to not worry too much about it -- if you could check that src/ci/docker/run.sh mingw-check works locally that would be sufficient for me (it may not actually detect deleted error codes but that's OK, just shouldn't fail).

**Comment by GuillaumeGomez** [comment]
```
$ src/ci/docker/run.sh mingw-check
...
Build completed successfully in 0:00:33
+ /scripts/validate-toolstate.sh
Cloning into 'rust-toolstate'...
<Nothing changed>
+ /scripts/validate-error-codes.sh
/scripts/validate-error-codes.sh: line 6: BASE_COMMIT: unbound variable
+ es-check es5 ../src/librustdoc/html/static/main.js ../src/librustdoc/html/static/search.js ../src/librustdoc/html/static/settings.js ../src/librustdoc/html/static/source-script.js ../src/librustdoc/html/static/storage.js
ES-Check: there were no ES version matching errors!  🎉
+ eslint ../src/librustdoc/html/static/main.js ../src/librustdoc/html/static/search.js ../src/librustdoc/html/static/settings.js ../src/librustdoc/html/static/source-script.js ../src/librustdoc/html/static/storage.js
Compile requests               237
Compile requests executed      215
Cache hits                      13
Cache misses                   202
Cache timeouts                   0
Cache read errors                0
Forced recaches                  0
Cache write errors               0
Compilation failures             0
Cache errors                     0
Non-cacheable compilations       0
Non-cacheable calls             18
Non-compilation calls            4
Unsupported compiler calls       0
Average cache write          0.000 s
Average cache read miss      0.105 s
Average cache read hit       0.000 s
Cache location             Local disk: "/sccache"
Cache size                     491 KiB
Max cache size                  10 GiB
== clock drift check ==
  local time: Mon Jun 28 19:08:09 UTC 2021
  network time: Mon, 28 Jun 2021 01:34:02 GMT
== end clock drift check ==
```

So this part is working fine at least.

**Comment by Mark-Simulacrum** [comment]
It looks like CI here is passing, but the E0001 code is deleted so presumably CI should fail. My guess is that the BASE_COMMIT variable is not set in the script -- you'll need to find a way to pipe it through. I'm not sure where it gets set, though, so there's likely some investigation that needs to be done.

File: src/ci/docker/host-x86_64/mingw-check/validate-error-codes.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# Checks that no error code explanation is removed.
+
+set -euo pipefail
+
+if (git log HEAD..."$BASE_COMMIT" --diff-filter=D --summary | grep delete \
```
Comment (bjorn3) [inline src/ci/docker/host-x86_64/mingw-check/validate-error-codes.sh:6]: Shouldn't the order be inverted?
```suggestion
if (git log "$BASE_COMMIT"...HEAD --diff-filter=D --summary | grep delete \
```

**Comment by GuillaumeGomez** [comment]
IT WORKS! :tada: 

Removing unwanted commit and it's good to go. :)

File: src/ci/scripts/should-skip-this.sh (hunk 1 above)
Comment (GuillaumeGomez) [reply src/ci/scripts/should-skip-this.sh]: Actually both syntaxes give the same result.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum This is now ready.

File: src/ci/docker/run.sh [hunk 4]
```diff
@@ -219,6 +219,10 @@ else
   command="/checkout/src/ci/run.sh"
 fi
 
+# Get some needed information for $BASE_COMMIT
+git fetch "https://github.com/$GITHUB_REPOSITORY" "$GITHUB_BASE_REF"
+BASE_COMMIT="$(git merge-base FETCH_HEAD HEAD)"
```
Comment (Mark-Simulacrum) [inline src/ci/docker/run.sh:224]: I think this is going to cause problems when trying to run these scripts locally -- GITHUB_BASE_REF and such are not well defined really. I think blocking this on `if [ "$CI" != "" ]; then` (like we do above in this file) and making sure that the validate-error-codes script only runs if BASE_COMMIT is defined.

**Comment by GuillaumeGomez** [comment]
Updated! Only setting `BASE_COMMIT` if on CI. Then the error code removal checker only runs if the variable is not empty (so if it doesn't exist, it doesn't run either).

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Should be ready now.

**Pull request merged**